### PR TITLE
Add C4Listener and JNI support

### PR DESCRIPTION
**Pull request by bmeike**
Sketch in C4Listener and its JNI support.

### CODE REVIEW

File: common/main/cpp/native_c4listener.cc [hunk 1]
```diff
@@ -0,0 +1,252 @@
+//
+// native_c4listener.cc
+//
+// Copyright (c) 2017 Couchbase, Inc All rights reserved.
```
Comment (jayahariv) [inline common/main/cpp/native_c4listener.cc:4]:
```suggestion
// Copyright (c) 2020 Couchbase, Inc All rights reserved.
```

File: common/main/cpp/native_c4listener.cc [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    config.networkInterface = hack; //networkInterface;
+    config.apis = apis;
+    config.tlsConfig = nullptr;
+    config.httpAuthCallback = httpAuthCallback;
+    config.callbackContext = (void *) context;
+    config.directory = hack; // dbPath;
+    config.allowCreateDBs = allowCreateDBs;
+    config.allowDeleteDBs = allowDeleteDBs;
+
+    auto listener = c4listener_start(&config, &error);
```
Comment (pasin) [inline common/main/cpp/native_c4listener.cc:168]: Throws if error occurs?

File: common/main/cpp/native_c4listener.cc [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    c4listener_getConnectionStatus(reinterpret_cast<C4Listener *>(c4Listener), &connections, &activeConnections);
+
+    return toConnectionStatus(env, connections, activeConnections);
+}
+
+JNIEXPORT jstring JNICALL
+Java_com_couchbase_lite_internal_core_C4Listener_getUriFromPath
+        (JNIEnv *env, jclass clazz, jlong c4Listener, jstring path) {
+    jstringSlice pathSlice(env, path);
+    auto uri = c4db_URINameFromPath(pathSlice);
```
Comment (pasin) [inline common/main/cpp/native_c4listener.cc:248]: IIRC, we may not need this.

File: common/main/java/com/couchbase/lite/internal/core/C4Listener.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        m.put(Option.REST, 0x01);
+        m.put(Option.SYNC, 0x02);
+        OPTION_TO_C4 = Collections.unmodifiableMap(m);
+    }
+
+    private static final Map<KeyMode, Integer> KEY_MODE_TO_C4;
+
+    static {
+        final Map<KeyMode, Integer> m = new HashMap<>();
+        m.put(KeyMode.CERT, 1);
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/internal/core/C4Listener.java:58]: For consistency, use hex like OPTION_TO_C4 or the other way around?

File: common/CMakeLists.txt [hunk 5]
```diff
@@ -126,8 +127,12 @@ add_dependencies(LiteCoreJNI JNIHeaders)
 add_subdirectory(${LITE_CORE_ROOT} ${CMAKE_LIBRARY_OUTPUT_DIRECTORY})
 
 target_link_libraries(  LiteCoreJNI
+                        "-Wl,--whole-archive"
                         LiteCoreStatic
+                        LiteCoreREST_Static
                         FleeceStatic
+                        "-Wl,--no-whole-archive"
```
[the remark this answers is not in the document]
Comment (bmeike) [reply common/CMakeLists.txt:134]: Yes.  It doesn't work wo/ that.  Seems like a linker bug to me.  Jim says it is SOP.

File: common/main/cpp/native_c4listener.cc (hunk 3 above)
Comment (bmeike) [reply common/main/cpp/native_c4listener.cc:248]: It is in the spec!!!

File: common/main/java/com/couchbase/lite/internal/core/C4Listener.java (hunk 4 above)
Comment (bmeike) [reply common/main/java/com/couchbase/lite/internal/core/C4Listener.java:58]: Oh... ok.  Sure.

Comment (bmeike) [reply common/main/java/com/couchbase/lite/internal/core/C4Listener.java:58]: Wait.  Options are bit flags.  They need to be hex.  Certs modes are not.  They are mutually exclusive.

These declarations are identical to the ones in c4Listener.cc

File: common/main/cpp/native_c4listener.cc (hunk 2 above)
Comment (bmeike) [reply common/main/cpp/native_c4listener.cc:168]: Yes.  Done.

File: common/main/cpp/native_c4listener.cc (hunk 1 above)
Comment (bmeike) [reply common/main/cpp/native_c4listener.cc]: fixed

**Pull request merged**